PonyGE2.wiki/Search-Options.md
# Search Loop and Step

## Search Loop

The search loop in PonyGE2 controls the overall duration of the search process, including the initialisation of the initial population, along with the main generations loop. In canonical GE, the search process loops over the total number of specified generations.

*__NOTE__ that in PonyGE2 the total number of generations refers to the number of generations over which the search process loops (i.e. over which evolution occurs), __NOT__ including initialisation. Thus, specifying 50 generations will mean an initial population will be generated and evaluated, and then the evolutionary process will loop for 50 generations. Since the initialised generation will be Generation 0, the total number of individuals evaluated across an entire evolutionary run will by __population x (generations + 1)__.*

The default search loop function in PonyGE2 is stored in `algorithm.search_loop`. However, users are free to implement their own search loops. These search loops can be stored either in the default `algorithm.search_loop` module, or wherever users desire. The search loop is a parameterisable function that can be [[set in the parameters dictionary|Evolutionary-Parameters#setting-parameters]].

While PonyGE2 is currently set up to only use the main search loop function (save for special cases such as re-loading an evolutionary run from state), it is possible for users to write their own search loop functions. It is possible to specify the desired search loop function directly through the parameters dictionary with the argument:

    --search_loop [SEARCH_LOOP]

or by setting the parameter `SEARCH_LOOP` to `[SEARCH_LOOP]` in either a parameters file or in the params dictionary, where `[SEARCH_LOOP]` is the name of the desired search loop function.

Along with the default search loop, a number of [[Hill Climbing|Search-Options#hill-climbing]] search loops are provided in PonyGE2.

## Step

At each generation in the main search loop, the `step` function is called, which by default is  `algorithm.step.step()`. The `step` function executes one full iteration of the canonical evolutionary process, typically:

1. [[Selection]]
2. [[Variation]]
   - [[Crossover|Variation#crossover]]
   - [[Mutation|Variation#mutation]]
3. [[Evaluation]]
4. [[Replacement]]

Step functions are stored in `algorithm.step`. While PonyGE2 is currently set up to only use the main step function, it is possible for users to write their own step functions. It is possible to specify the desired step function directly through the parameters dictionary with the argument:

    --step [STEP]

or by or by setting the parameter `STEP` to `[STEP]` in either a parameters file or in the params dictionary, where `[STEP]` is the name of the desired step function.

*__NOTE__ that if a step function is saved in the main* `algorithm.step` *module, there is no need to specify the full path to the function. However, if the desired step function is located elsewhere, it __is__ necessary to specify the full module path to the function.*


# Population Options

There are a number of parameters within PonyGE2 for controlling overall populations.

## Population Size

The population size controls the total number of individuals to be generated at each generation. The default value is 500. This value can be changed with the argument:

    --population_size [INT]

or by setting the parameter `POPULATION_SIZE` to `[INT]` in either a parameters file or in the params dictionary, where `[INT]` is an integer which specifies the population size.

Higher population sizes can improve performance on difficult problems, but require more computational effort and may lead to premature convergence.

## Generations

The number of generations the evolutionary algorithm will run for. The default value is 50. This value can be changed with the argument:

    --generations [INT]

or by setting the parameter `GENERATIONS` to `[INT]` in either a parameters file or in the params dictionary, where `[INT]` is an integer which specifies the number of generations.

Higher numbers of generations can improve performance, but will lead to longer run-times.


# Hill Climbing

A simple hill-climbing approach, and two modern variants, are provided in `algorithm.hill_climbing`. In simple HC, there is a single current individual (i.e. no population). At each step of the search loop, a mutation operator creates a new candidate individual. If this individual is better than the original, the move is "accepted" -- the candidate replaces the current individual. In Late-Acceptance Hill-Climbing (LAHC), due to [Bykov](http://www.cs.nott.ac.uk/~yxb/LAHC/LAHC-TR.pdf), and Step-Counting Hill-Climbing (SCHC), due to [Bykov](http://link.springer.com/article/10.1007/s10951-016-0469-x), the acceptance decision depends on whether the candidate is better than one considered some time previously, dependent on a "history length" parameter. This can allow search to escape local optima. The following arguments will select one of these methods:

    --search_loop algorithm.hill_climbing.LAHC_search_loop

    --search_loop algorithm.hill_climbing.SCHC_search_loop

or by setting the parameter `SEARCH_LOOP` to either `algorithm.hill_climbing.LAHC_search_loop` or `algorithm.hill_climbing.SCHC_search_loop` in either a parameters file or the parameters dictionary. 

The "history length" parameter is set as follows:

    --hill_climbing_history [INT]

or by setting the parameter `HILL_CLIMBING_HISTORY` to `[INT]` in either a parameters file or the parameters dictionary, where `[INT]` is an integer value specifying the desired hill climbing history length.

Step-Counting Hill Climbing has an additional `SCHC_COUNT_METHOD` parameter that can be specified. This parameter sets the counting method to be used with the SCHC algorithm. This parameter can be set with the argument:

    --schc_count_method [METHOD]

or by setting the parameter `SCHC_COUNT_METHOD` to `[METHOD]` in either a parameters file or the parameters dictionary, where `[METHOD]` is a string specifying the desired step counting method. Currently implemented acceptable strings are:
 
1. "count_all"
    
    This is the default option, and counts all moves.
2. "acp"

    This option counts accepted moves only (i.e. where the candidate solution is better than the cost bound and better than or equal to the current best).
3. "imp"

    This option counts improving moves only (i.e. where the candidate solution is better than the current best).

*__NOTE__ that both LAHC and SCHC reduce to a simple hill-climbing algorithm when used with a history length of 1.*

*__NOTE__ that crossover is __not__ used in hill-climbing algorithms.*


# Distributed Search

In the standard implementation of PonyGE2, the evolutionary search process runs on a single machine (or agent) which has access to all the individuals in the population. It can then choose among the highest performing individuals from the population to perform the various genetic operations. In a multi-agent implementation, each agent has its own instance of genetic information. An agent can only have access to its own genotype. The agent moves in an environment and communicates to nearby agents to share the genetic information. After collecting genetic information from a handful of agents, it can apply the genetic operations by itself for the information that it has gathered. This multi-agent approach is particularly useful for robotics or distributed implementations across many nodes, as each agent can act individually with only local knowledge.

## Distributed Search Loop

When the `MULTIAGENT` parameter is set to `True`, search_loop and step methods from `algorithm.distributed_algorithm` automatically get activated. The first step in the distributed search loop is to initialize the various agents objects. Based on the parameter set for `AGENT_SIZE` that the number of agents is created, where the `INTERACTION_PROBABILITY` parameter is passed while initializing agents. This parameter defines the probability of interaction with other agents. Higher this value, there are higher changes interacting with other agents and higher chances of finding solutions faster. Then after initializing agents, for each generation, step module is called which takes the list of agents as a function parameter and returns a list of agents.

## Distributed Step

In this module, for each agent in the environment, the agent will perform: sense, act and update action sequentially. These three action or modules are built in for the agents. These modules can be extended or overridden based on the type of agent required to solve a particular problem. After performing these actions, a list of agents is returned.

## Agents

It is the class that defines the behavior of the agents. Right now we have only defined a very primitive agent but based on the requirements the method of this class can be overridden. The Sense method describes the way the agent gets information from the environment. The Act method provides a way from the agent to process the information obtained from Sense method. The update method is agent's way to update its state. In our case, the Sense method is used to find other nearby agents, the Act method is used to apply genetic operations to the stack of genetic information collected from nearby agents and the update method is used to update the agent's genetic information.


PonyGE2.wiki/Evolutionary-Parameters.md
One of the central components of PonyGE is the `algorithm.parameters.params` dictionary. This dictionary is referenced throughout the entire program and is used to streamline the whole process by keeping all optional parameters in the one place. This also means that there is little to no need for arguments to the various functions in PonyGE, as these arguments can often be read directly from the parameters dictionary. Furthermore, the parameters dictionary is used to specify and store optional functions such as `initialisation`, `crossover`, `mutation`, and `replacement`.

# Setting Parameters

There are three different ways to specify operational parameters with PonyGE: by editing the dictionary itself, by setting a parameters file, and by specifying parameters from the command line.

## The Parameters Dictionary

The most basic way to set specific parameters values in PonyGE2 is to directly modify the `algorithm.parameters.params` dictionary in the code. This is not encouraged, as the `algorithm.parameters.params` dictionary contains the default values for many parameters. As such, PonyGE2 may perform in an unpredictable manner in some cases.

## Parameters Files

The most versatile and robust method for specifying desired parameters is to list your desired parameters in a specialised parameters text file. Specific parameters files can then be read in by PonyGE2, and will set any parameters specified therein.

Parameters files are simple text files (with no commenting or other documentation) that simply list all of the desired parameters to be set in the central `algorithm.parameters.params` dictionary. Parameters can be specified in a parameters file using the following formatting:
 
    PARAMETER_NAME: PARAMETER_VALUE
    
where `PARAMETER_NAME` is the name of the desired parameter from the central `algorithm.parameters.params` dictionary to be set, and `PARAMETER_VALUE` is the desired value to be stored for that parameter.

Entire lines can be commented out of parameters files simply by prefacing the line with a `#` symbol, as with normal Python code. The use of comments allows parameters files to be neatly and clearly organised into grouped sections if so desired. Note that blank lines in parameters files are ignored.

*__NOTE__ that parameter names and values must be colon-separated, i.e. each line of the parameters file will be read in as a string, and the string will be parsed based on the location of the first instance of a colon. __NOTE__ that this __does__ preclude the use of colons in parameter __names__ (i.e. the keys of the params dictionary), but that this does __not__ preclude the use of colons in the parameter __values__.*

Example parameters files are located in the `parameters` folder. When using parameters files, it is necessary to specify the desired parameter file from the command line. This is done with the argument:

    --parameters [FULL FILE NAME INCLUDING EXTENSION]

## Command-Line Arguments

The third and final method for setting parameters is to list desired parameters directly from the command line. PonyGE2 has an extensive command line parser, located in `utilities.algorithm.command_line_parser`. All stock parameters listed in the original `algorithm.parameters.params` dictionary can be set or changed via command line arguments. Command line arguments for all parameters are the same as the dictionary keys, except all in lower case. For example, the dictionary key for the size of the evolutionary population is `POPULATION_SIZE`. This value can be set from the command line with the argument:

    --population_size [SIZE]
    
where `[SIZE]` is an integer specifying the desired population size.

To see a list of all currently available command-line arguments implemented in the parser, type

    $ python ponyge.py --help

## Order of Setting Parameters

Parameters are set in the `algorithm.parameters.params` dictionary in strict order based on the above three methods. Initial values are set in directly by the [[dictionary|Evolutionary-Parameters#the-parameters-dictionary]] itself. Any specified [[parameters file|Evolutionary-Parameters#parameters-files]] is then read in, and any values specified therein will over-write the values contained so far in the dictionary. Finally, any specified [[command-line arguments|Evolutionary-Parameters#command-line-arguments]] will over-write all previously set parameters. 

## Path Specification

The modular structure of PonyGE2 has specialised modules (folders) for each aspect of the overall system:

1. [[Datasets|Evaluation#datasets]] *must* be stored in the `datasets` folder,
2. [[Grammar files|Grammars]] *must* be stored in the `grammars` folder,
3. [[Parameters files|Evolutionary-Parameters#parameters-files]] *must* be stored in the `parameters` folder,
4. [[Population seeds|Seeding#2-seeding-runs-with-one-or-more-target-solutions]] *must* be stored in the `seeds` folder,
5. [[Fitness functions|Evaluation#fitness-functions]] *must* be stored in `src.fitness`,
6. [[Error metrics|Evaluation#error-metrics]] *must* be stored in `utilities.fitness.error_metric`,
7. Operators are typically stored in the operators folder:
    1. [[Initialisation]] functions are stored in `src.operators.initialisation`,
    2. [[Selection]] functions are stored in `src.operators.selection`,
    3. [[Crossover|Variation#crossover]] functions are stored in `src.operators.crossover`,
    4. [[Mutation|Variation#mutation]] functions are stored in `src.operators.mutation`,
    5. [[Replacement]] functions are stored in `src.operators.replacement`.
8. [[Search framework functions|Search-Options]] are typically stored in the algorithm folder:
    1. [[Search Loop|Search-Options#search-loop]] functions are typically stored in `src.algorithm.search_loop`,
    2. [[Step|Search-Options#step]] functions are typically stored in `src.algorithm.step`.

Numbers 1-6 above are hard-coded into the PonyGE2 framework, i.e. these various components of PonyGE2 *must* be stored in their respective folders. However, optional operators have more freedom in the PonyGE2 framework (this will be explained presently).

For all specified arguments listed above, PonyGE2 automatically parses the correct path. Since numbers 1-6 above are hard-coded, only the direct names of the desired modules or functions within the respective folders/modules need to be specified. For example, when specifying a grammar file, one simply needs to specify the path to the file itself *from within the `grammars` folder*, i.e. `letter.bnf` or `supervised_learning/Dow.bnf`. When specifying a fitness function, one simply needs to specify the name of the function class itself *from within the `fitness` module*, i.e. `string_match` or `supervised_learning.regression`. There is no need to specify the containing folder or module (e.g. `grammars` or `fitness` in the above two cases, respectively). Doing so will produce an error.

However, operators and search framework functions (as listed in points 7 and 8 above) have more freedom in where they can be saved. These functions can either be located in their respective default `operators` or `algorithm` modules, or in any desired location within the overall PonyGE2 `src` directory. 

If you are importing operators that are saved in their respective default locations, there is no need to specify the full module path; you can simply just specify the name of the desired operator function. For example, if you are seeking to use PonyGE2's built-in [[subtree crossover|Variation#subtree]] function, then you simply need to specify the function name itself, `subtree`, for the relevant parameter input (e.g. `CROSSOVER: subtree` in a parameters file, or `--crossover subtree` from the command line). To summarize, if a single function name is specified, PonyGE2 will attempt to load that function from its default location. 

On the other hand, you can specify full module paths (from within the `src` directory) to operator or search functions should you so desire. This enables if you to import these methods from any location within the `src` directory. For example, if you are seeking to use PonyGE2's built-in [[Late-Acceptance Hill Climbing|Search-Options#hill-climbing]] function, you can specify `algorithm.hill_clilmbing.LAHC_search_loop` for the relevant parameter input (as above). PonyGE2 will attempt to import the module directly from the full specified path, and if this fails, it will then try to import the module from the `algorithm` module location (taking the specified path into account).

If PonyGE2 cannot find the specified functions or modules, error reports will be printed out specifying exactly where PonyGE2 is trying to search. Should they appear, these reports should help you to diagnose what has gone awry.

# Adding New Parameters
-----------------------

It is possible to add new parameters to the `algorithm.parameters.params` dictionary either by editing the dictionary directly (this is not recommended), or simply by specifying any desired parameters in a parameters file. Since parameters files are used to set new or over-write existing key:value pairs in the parameters dictionary, it is possible to specify any number of new/unseen parameters in a parameters file (even if they do not already appear in the default `algorithm.parameters.params` dictionary). Such new or custom parameters can then be accessed throughout the entire PonyGE2 code base. However, it is not currently possible to specify new parameters directly from the command line. This is because any existing parameters that were merely mis-spelled from the command line would then create new parameters, and the evolutionary system would not perform as expected. Thus, any parameters entered from the command line that do not exist in the command line parser in `utilities.algorithm.command_line_parser` will produce an error. Of course, it is entirely possible to modify the command line parser to accept new arguments.

*__NOTE__ that parameter names should follow the naming convention demonstrated with the existing parameters. All names should contain only uppercase letters and underscores. Avoid the use of spaces where possible and do not use a colon (`:`) in the name of a parameter.*


PonyGE2/src/algorithm/parameters.py
from multiprocessing import cpu_count
from os import path
from socket import gethostname

hostname = gethostname().split('.')
machine_name = hostname[0]

"""Algorithm parameters"""
params = {
    # Set default step and search loop functions
    'SEARCH_LOOP': 'search_loop',
    'STEP': 'step',

    # Evolutionary Parameters
    'POPULATION_SIZE': 500,
    'GENERATIONS': 50,
    'HILL_CLIMBING_HISTORY': 1000,
    'SCHC_COUNT_METHOD': "count_all",

    # Set optional experiment name
    'EXPERIMENT_NAME': None,
    # Set default number of runs to be done.
    # ONLY USED WITH EXPERIMENT MANAGER.
    'RUNS': 1,

    # Class of problem
    'FITNESS_FUNCTION': "supervised_learning.regression",

    # Select problem dataset
    'DATASET_TRAIN': "Vladislavleva4/Train.txt",
    'DATASET_TEST': None,
    'DATASET_DELIMITER': None,

    # Set grammar file
    'GRAMMAR_FILE': "supervised_learning/Vladislavleva4.bnf",

    # Set the number of depths permutations are calculated for
    # (starting from the minimum path of the grammar).
    # Mainly for use with the grammar analyser script.
    'PERMUTATION_RAMPS': 5,

    # Select error metric
    'ERROR_METRIC': None,

    # Optimise constants in the supervised_learning fitness function.
    'OPTIMIZE_CONSTANTS': False,

    # Specify target for target problems
    'TARGET': "ponyge_rocks",

    # Set max sizes of individuals
    'MAX_TREE_DEPTH': 90,  # SET TO 90 DUE TO PYTHON EVAL() STACK LIMIT.
    # INCREASE AT YOUR OWN RISK.
    'MAX_TREE_NODES': None,
    'CODON_SIZE': 100000,
    'MAX_GENOME_LENGTH': None,
    'MAX_WRAPS': 0,

    # INITIALISATION
    # Set initialisation operator.
    'INITIALISATION': "operators.initialisation.PI_grow",
    # Set the maximum genome length for initialisation.
    'INIT_GENOME_LENGTH': 200,
    # Set the maximum tree depth for initialisation.
    'MAX_INIT_TREE_DEPTH': 10,
    # Set the minimum tree depth for initialisation.
    'MIN_INIT_TREE_DEPTH': None,

    # SELECTION
    # Set selection operator.
    'SELECTION': "operators.selection.tournament",
    # For tournament selection
    'TOURNAMENT_SIZE': 2,
    # For truncation selection
    'SELECTION_PROPORTION': 0.5,
    # Allow for selection of invalid individuals during selection process.
    'INVALID_SELECTION': False,

    # OPERATOR OPTIONS
    # Boolean flag for selecting whether or not mutation is confined to
    # within the used portion of the genome. Default set to True.
    'WITHIN_USED': True,

    # CROSSOVER
    # Set crossover operator.
    'CROSSOVER': "operators.crossover.variable_onepoint",
    # Set crossover probability.
    'CROSSOVER_PROBABILITY': 0.75,
    # Prevents crossover from generating invalids.
    'NO_CROSSOVER_INVALIDS': False,

    # MUTATION
    # Set mutation operator.
    'MUTATION': "operators.mutation.int_flip_per_codon",
    # Set mutation probability (None defaults to 1 over the length of
    # the genome for each codon)
    'MUTATION_PROBABILITY': None,
    # Set number of mutation events
    'MUTATION_EVENTS': 1,
    # Prevents mutation from generating invalids.
    'NO_MUTATION_INVALIDS': False,

    # REPLACEMENT
    # Set replacement operator.
    'REPLACEMENT': "operators.replacement.generational",
    # Set elite size.
    'ELITE_SIZE': None,

    # DEBUGGING
    # Use this to turn on debugging mode. This mode doesn't write any files
    # and should be used when you want to test new methods.
    'DEBUG': False,

    # PRINTING
    # Use this to print out basic statistics for each generation to the
    # command line.
    'VERBOSE': False,
    # Use this to prevent anything being printed to the command line.
    'SILENT': False,

    # SAVING
    # Save the phenotype of the best individual from each generation. Can
    # generate a lot of files. DEBUG must be False.
    'SAVE_ALL': False,
    # Save a plot of the evolution of the best fitness result for each
    # generation.
    'SAVE_PLOTS': True,

    # MULTIPROCESSING
    # Multi-core parallel processing of phenotype evaluations.
    'MULTICORE': False,
    # Set the number of cpus to be used for multiprocessing
    'CORES': cpu_count(),

    # STATE SAVING/LOADING
    # Save the state of the evolutionary run every generation. You can
    # specify how often you want to save the state with SAVE_STATE_STEP.
    'SAVE_STATE': False,
    # Specify how often the state of the current evolutionary run is
    # saved (i.e. every n-th generation). Requires int value.
    'SAVE_STATE_STEP': 1,
    # Load an evolutionary run from a saved state. You must specify the
    # full file path to the desired state file. Note that state files have
    # no file type.
    'LOAD_STATE': None,

    # SEEDING
    # Specify a list of PonyGE2 individuals with which to seed the initial
    # population.
    'SEED_INDIVIDUALS': [],
    # Specify a target seed folder in the 'seeds' directory that contains a
    # population of individuals with which to seed a run.
    'TARGET_SEED_FOLDER': None,
    # Set a target phenotype string for reverse mapping into a GE
    # individual
    'REVERSE_MAPPING_TARGET': None,
    # Set Random Seed for all Random Number Generators to be used by
    # PonyGE2, including the standard Python RNG and the NumPy RNG.
    'RANDOM_SEED': None,

    # CACHING
    # The cache tracks unique individuals across evolution by saving a
    # string of each phenotype in a big list of all phenotypes. Saves all
    # fitness information on each individual. Gives you an idea of how much
    # repetition is in standard GE/GP.
    'CACHE': False,
    # Uses the cache to look up the fitness of duplicate individuals. CACHE
    # must be set to True if you want to use this.
    'LOOKUP_FITNESS': False,
    # Uses the cache to give a bad fitness to duplicate individuals. CACHE
    # must be True if you want to use this (obviously)
    'LOOKUP_BAD_FITNESS': False,
    # Removes duplicate individuals from the population by replacing them
    # with mutated versions of the original individual. Hopefully this will
    # encourage diversity in the population.
    'MUTATE_DUPLICATES': False,

    # MULTI-AGENT Parameters
    # True or False for multi-agent
    'MULTIAGENT': False,
    # Agent Size. Number of agents having their own copy of genetic material
    'AGENT_SIZE': 100,
    # Interaction Probability: how frequently the agents can interaction with
    # each other
    'INTERACTION_PROBABILITY': 0.5,

    # OTHER
    # Set machine name (useful for doing multiple runs)
    'MACHINE': machine_name
}


def load_params(file_name):
    """
    Load in a params text file and set the params dictionary directly.

    :param file_name: The name/location of a parameters file.
    :return: Nothing.
    """

    try:
        open(file_name, "r")
    except FileNotFoundError:
        s = "algorithm.parameters.load_params\n" \
            "Error: Parameters file not found.\n" \
            "       Ensure file extension is specified, e.g. 'regression.txt'."
        raise Exception(s)

    with open(file_name, 'r') as parameters:
        # Read the whole parameters file.
        content = parameters.readlines()

        for line in [l for l in content if not l.startswith("#")]:

            # Parameters files are parsed by finding the first instance of a
            # colon.
            split = line.find(":")

            # Everything to the left of the colon is the parameter key,
            # everything to the right is the parameter value.
            key, value = line[:split], line[split + 1:].strip()

            # Evaluate parameters.
            try:
                value = eval(value)

            except:
                # We can't evaluate, leave value as a string.
                pass

            # Set parameter
            params[key] = value


def set_params(command_line_args, create_files=True):
    """
    This function parses all command line arguments specified by the user.
    If certain parameters are not set then defaults are used (e.g. random
    seeds, elite size). Sets the correct imports given command line
    arguments. Sets correct grammar file and fitness function. Also
    initialises save folders and tracker lists in utilities.trackers.

    :param command_line_args: Command line arguments specified by the user.
    :return: Nothing.
    """

    from utilities.algorithm.initialise_run import initialise_run_params
    from utilities.algorithm.initialise_run import set_param_imports
    from utilities.fitness.math_functions import return_one_percent
    from utilities.algorithm.command_line_parser import parse_cmd_args
    from utilities.stats import trackers, clean_stats
    from representation import grammar

    cmd_args, unknown = parse_cmd_args(command_line_args)

    if unknown:
        # We currently do not parse unknown parameters. Raise error.
        s = "algorithm.parameters.set_params\nError: " \
            "unknown parameters: %s\nYou may wish to check the spelling, " \
            "add code to recognise this parameter, or use " \
            "--extra_parameters" % str(unknown)
        raise Exception(s)

    # LOAD PARAMETERS FILE
    # NOTE that the parameters file overwrites all previously set parameters.
    if 'PARAMETERS' in cmd_args:
        load_params(path.join("..", "parameters", cmd_args['PARAMETERS']))

    # Join original params dictionary with command line specified arguments.
    # NOTE that command line arguments overwrite all previously set parameters.
    params.update(cmd_args)

    if params['LOAD_STATE']:
        # Load run from state.
        from utilities.algorithm.state import load_state

        # Load in state information.
        individuals = load_state(params['LOAD_STATE'])

        # Set correct search loop.
        from algorithm.search_loop import search_loop_from_state
        params['SEARCH_LOOP'] = search_loop_from_state

        # Set population.
        setattr(trackers, "state_individuals", individuals)

    else:
        if params['REPLACEMENT'].split(".")[-1] == "steady_state":
            # Set steady state step and replacement.
            params['STEP'] = "steady_state_step"
            params['GENERATION_SIZE'] = 2

        else:
            # Elite size is set to either 1 or 1% of the population size,
            # whichever is bigger if no elite size is previously set.
            if params['ELITE_SIZE'] is None:
                params['ELITE_SIZE'] = return_one_percent(1, params[
                    'POPULATION_SIZE'])

            # Set the size of a generation
            params['GENERATION_SIZE'] = params['POPULATION_SIZE'] - \
                                        params['ELITE_SIZE']

        if (params['MUTATION_PROBABILITY'] is not None and
            params['MUTATION_EVENTS'] != 1):
            s = "operators.mutation.int_flip_per_codon\n" \
                "Error: mutually exclusive parameters 'MUTATION_PROBABILITY'" \
                "and 'MUTATION_EVENTS' have been explicitly set.\n" \
                "Only one of these parameters can be used at a time."
            raise Exception(s)

        # Initialise run lists and folders before we set imports.r
        initialise_run_params(create_files)

        # Set correct param imports for specified function options, including
        # error metrics and fitness functions.
        set_param_imports()

        # Clean the stats dict to remove unused stats.
        clean_stats.clean_stats()

        # Set GENOME_OPERATIONS automatically for faster linear operations.
        if (params['CROSSOVER'].representation == "subtree" or
                params['MUTATION'].representation == "subtree"):
            params['GENOME_OPERATIONS'] = False
        else:
            params['GENOME_OPERATIONS'] = True

        # Ensure correct operators are used if multiple fitness functions used.
        if hasattr(params['FITNESS_FUNCTION'], 'multi_objective'):

            # Check that multi-objective compatible selection is specified.
            if not hasattr(params['SELECTION'], "multi_objective"):
                s = "algorithm.parameters.set_params\n" \
                    "Error: multi-objective compatible selection " \
                    "operator not specified for use with multiple " \
                    "fitness functions."
                raise Exception(s)

            if not hasattr(params['REPLACEMENT'], "multi_objective"):

                # Check that multi-objective compatible replacement is
                # specified.
                if not hasattr(params['REPLACEMENT'], "multi_objective"):
                    s = "algorithm.parameters.set_params\n" \
                        "Error: multi-objective compatible replacement " \
                        "operator not specified for use with multiple " \
                        "fitness functions."
                    raise Exception(s)

        # Parse grammar file and set grammar class.
        params['BNF_GRAMMAR'] = grammar.Grammar(
            path.join("..", "grammars", params['GRAMMAR_FILE']))

        # If OPTIMIZE_CONSTANTS, check that the grammar is suitable
        if params['OPTIMIZE_CONSTANTS']:
            if "c[" not in params['BNF_GRAMMAR'].terminals:
                raise ValueError("Grammar unsuitable for OPTIMIZE_CONSTANTS")

        # Population loading for seeding runs (if specified)
        if params['TARGET_SEED_FOLDER']:

            # Import population loading function.
            from operators.initialisation import load_population

            # A target folder containing seed individuals has been given.
            params['SEED_INDIVIDUALS'] = load_population(
                params['TARGET_SEED_FOLDER'])

        elif params['REVERSE_MAPPING_TARGET']:
            # A single seed phenotype has been given. Parse and run.

            # Import GE LR Parser.
            from scripts import GE_LR_parser

            # Parse seed individual and store in params.
            params['SEED_INDIVIDUALS'] = [GE_LR_parser.main()]
